### ISSUE TITLE
Add @Nullable @NonNull throught the library

### ISSUE THREAD

**Issue by PaulWoitaschek** [square/moshi#223]
It would be great to have the whole api annotated with `@Nullable` and `@NonNull`.
Also it would play way better together with kotlin. For example the JsonAdapter.Factory:

```java
    /**
     * Attempts to create an adapter for {@code type} annotated with {@code annotations}. This
     * returns the adapter if one was created, or null if this factory isn't capable of creating
     * such an adapter.
     *
     * <p>Implementations may use to {@link Moshi#adapter} to compose adapters of other types, or
     * {@link Moshi#nextAdapter} to delegate to the underlying adapter of the same type.
     */
    JsonAdapter<?> create(Type type, Set<? extends Annotation> annotations, Moshi moshi);
```

Currently one really has to read the documentation.
What kotlin does by default is:
```kotlin
override fun create(type: Type?, annotations: MutableSet<out Annotation>?, moshi: Moshi?): JsonAdapter<*> 
```
where it should be
```kotlin
override fun create(type: Type, annotations: MutableSet<out Annotation>, moshi: Moshi): JsonAdapter<*>?
```

**Comment by NightlyNexus** [comment]
Maybe there should be one big issue on one project somewhere to communicate plans?
kind of tracking this here: https://github.com/square/retrofit/issues/624#issuecomment-265429436

**Comment by PaulWoitaschek** [comment]
Something else: Validation of null input is not done Properly. I just received a bit weird crash because I passted null to `JsonAdapter.fromJson(String)` and the `Buffer` called `string.length()` in `writeUtf8` and caused a `NPE`. It would be handy if you could check the arguments and fail immediately rather than relying on the implementation to crash.

**Comment by swankjesse** [comment]
Implemented. https://github.com/square/moshi/pull/297

**Comment by PaulWoitaschek** [comment]
If I understand correctly the String is still [not annotated](https://github.com/square/moshi/blob/master/moshi/src/main/java/com/squareup/moshi/JsonAdapter.java#L39) and will crash in Okios Buffer?

**Comment by joshfriend** [comment]
[`package-info.java` is annotated with `@ParametersAreNonnullByDefault`](https://github.com/square/moshi/blob/de4c2e782e0979e8b747474ce45585821c15e690/moshi/src/main/java/com/squareup/moshi/package-info.java#L2) but Kotlin [doesn't support that yet.](https://youtrack.jetbrains.com/issue/KT-10942)

**Comment by PaulWoitaschek** [comment]
Annotated or not, it's still crashing because a dependent is crashing.

**Comment by swankjesse** [comment]
If this is blocking you, please vote this up: https://youtrack.jetbrains.com/oauth?state=%2Fissue%2FKT-10942

**Comment by 3flex** [comment]
> package-info.java is annotated with @ParametersAreNonnullByDefault but Kotlin doesn't support that yet.

Should be fixed in Kotlin 1.1.4. Support for the annotations can be enabled by passing the `-Xjsr305-annotations=enable` command line option to the compiler. Info in the [release announcement](https://blog.jetbrains.com/kotlin/2017/08/kotlin-1-1-4-is-out/).

**Comment by swankjesse** [comment]
No further action for Moshi on this.
